### ISSUE TITLE
Feature request: Terraform Parameter support

### ISSUE THREAD

**Issue by bashtoni** [crayfishx/jerakia#116]
It would be great to have a hierarchy of parameters to pass to Terraform for a plan/apply.

The use case is that often we will have a number of test environments created by Terraform.  Most of these have parameters that are mostly the same, but with a couple of differences.

We're currently solving this with multiple parameter files, but this doesn't provide the inherited hierarchy we'd like.

**Comment by beddari** [comment]
Yes, this is something we'd like to work on. @jtopjian mentioned at Cfmgmtcamp he could be interested in writing a provider, and my own team would also like this.

Thanks for prodding 👍

**Comment by jtopjian** [comment]
Yep, I'm definitely interested. I hope to have some time in the next week to start looking into it. 

A rough outline would be:

1. Learn the Jerakia API.
2. Create a rough Go library/package to interact with the Jerakia API.
3. Write a Terraform data source (read-only) and maybe a resource (read/write).

**Comment by beddari** [comment]
Yes, that was about what I was thinking too. It would be tremendously helpful to at least get a skeleton from someone with more golang exp :) And I'm hoping to attract some of the mgmt crowd eventually :) /cc @purpleidea

**Comment by purpleidea** [comment]
@beddari We in https://github.com/purpleidea/mgmt/ are happy to be down with jerakia, and we're waiting on a longer discussion about if we can get events when data changes basically... A golang implementation would also help :)

**Comment by jtopjian** [comment]
I've put together a basic Go client here: https://github.com/jtopjian/go-jerakia

**Comment by beddari** [comment]
This is awesome, thanks! I'll take a look at it over the next few days.

Also I think it'd be good to create some client integration/acceptance tests, so that each lib implementation can handle edge cases in a similar way. E.g as an example, setting a namespace is optional, not required. Actually not sure what the docs say about that ;-) so having tests would help.

**Comment by jtopjian** [comment]
> Also I think it'd be good to create some client integration/acceptance tests, so that each lib implementation can handle edge cases in a similar way.

I agree.

Rather than creating my own fixtures, I used the ones bundled with Jerakia as a form of standardization. I can search the Jerakia code for the built-in unit tests and convert them to Go. The ones in-place at the moment are purely to test a proof-of-concept.

> setting a namespace is optional, not required

At least with the REST API, namespace is required:

```json
{"status":"failed","message":"Must include parameter namespace in request"}
```

**Comment by crayfishx** [comment]
@jtopjian Firstly thanks for your work on this - it's awesome!

Namespaces were initially a required thing, but as we looked at other scenarios and tooling I realized that this might be a bit restrictive and to maybe make it up to the data source to decide whether or not to enforce a namespace, therefore it was relaxed in the CLI and the ruby API to be optional.... the API does still enforce this, as you've noted - I'm happy to relax this and remove the restriction. This can be done as a non breaking change I think

**Comment by beddari** [comment]
@crayfishx Yes, please remove that req, I'd also like to optionally not set namespace from the Python client. Should the current namespace is an empty string behavior be kept, then? I think it would be good to make empty string an error, later on, or ...?

**Comment by crayfishx** [comment]
@beddari @jtopjian This is handled in #117

**Comment by jtopjian** [comment]
@crayfishx Nice! I will update `go-jerakia` when I have a moment.

**Comment by jtopjian** [comment]
wow - where did the year go? :)

I have a basic Terraform provider created and can publish it once https://github.com/jerakia/go-jerakia/pull/3 is taken care of.

**Comment by jtopjian** [comment]
I've published https://github.com/jtopjian/terraform-provider-jerakia for review. If/Once it looks good, we can move it to the main Jerakia GitHub org (I think I have permissions to create new repositories, but didn't want to create a repo without prior discussion).

There's a quirk with how results are handled:

Jerakia can return all sorts of results (strings, bools, arrays, hashes) but Terraform uses a strict schema for attributes. These two don't mix well together.

For now, I have created an attribute called `result_json` which will contain the result of the lookup encoded as JSON and then encoded as a string. This can then be used as input for other resources/actions where those resources/actions can decode/parse the string as a proper JSON structure.

This is a common Terraform pattern for these situations, although it's not the best experience.

There are two things we can do:

1. Leave as-is

2. Keep `result_json` and add other attributes such as `result_string`, `result_bool`, `result_int`. These attributes will be set when the result of the lookup is a single item that has been detected to be one of those types. Anything else (array, hash) will have to go in `result_json`. Terraform has support for Lists and Map attributes, but it's possible the array/list contains sub-arrays and sub-hashes.
    
    The advantage of this solution is that it simplifies parsing of results when you are absolutely confident that the lookup will return a single string item. The disadvantage is when you don't know. We'll have to put the responsibility on the user to make that determination and use the appropriate `result_*` attribute.

Thoughts?

**Comment by beddari** [comment]
My vote is for "leave as-is" ... and document how this works. I've seen examples with other custom data providers but it's been a while since I last touched that so I'm sure you have better ones Joe.
